File: code-tiny/tools/spring/value_resolver.py
from __future__ import annotations

import re
from typing import Any, Dict, Iterable, List, Optional, Sequence, Tuple


_STRING_RE = re.compile(r'"([^"\\]*(?:\\.[^"\\]*)*)"|\'([^\'\\]*(?:\\.[^\'\\]*)*)\'')
_NAMED_ARG_RE = re.compile(r"([A-Za-z_][\w.-]*)\s*=\s*([^,]+(?:\{[^}]*\})?)")
_PLACEHOLDER_RE = re.compile(r"\$\{([^}:]+)(?::([^}]+))?\}")


def strip_quotes(value: str) -> str:
    text = (value or "").strip()
    if len(text) >= 2 and text[0] == text[-1] and text[0] in {"'", '"'}:
        return text[1:-1]
    return text


def extract_string_literals(text: str) -> List[str]:
    values: List[str] = []
    for match in _STRING_RE.finditer(text or ""):
        values.append(match.group(1) if match.group(1) is not None else match.group(2))
    return values


def parse_annotation_args(raw_args: str) -> Dict[str, Any]:
    text = (raw_args or "").strip()
    if text.startswith("(") and text.endswith(")"):
        text = text[1:-1].strip()
    result: Dict[str, Any] = {}
    if not text:
        return result
    for key, value in _NAMED_ARG_RE.findall(text):
        result[key] = parse_value(value)
    if not result:
        result["value"] = parse_value(text)
    return result


def parse_value(value: str) -> Any:
    text = (value or "").strip()
    if text.startswith("{") and text.endswith("}"):
        return [parse_value(item) for item in _split_top_level(text[1:-1])]
    strings = extract_string_literals(text)
    if len(strings) == 1 and text in {f'"{strings[0]}"', f"'{strings[0]}'"}:
        return strings[0]
    if strings and text.startswith("{"):
        return strings
    if text.lower() in {"true", "false"}:
        return text.lower() == "true"
    return strip_quotes(text)


def first_arg(args: Dict[str, Any], *names: str, default: str = "") -> Any:
    for name in names:
        if name in args:
            return args[name]
    return default


def list_arg(args: Dict[str, Any], *names: str) -> List[str]:
    value: Any = first_arg(args, *names, default=[])
    if value is None or value == "":
        return []
    if isinstance(value, list):
        return [str(item) for item in value if str(item)]
    return [str(value)]


def resolve_placeholders(value: str, config_index: Dict[str, List[Any]]) -> Tuple[str, str]:
    raw = str(value or "")
    status = "resolved"

    def repl(match: re.Match[str]) -> str:
        nonlocal status
        key = match.group(1)
        default = match.group(2)
        candidates = config_index.get(key) or []
        if candidates:
            return str(candidates[-1])
        if default is not None:
            status = "defaulted"
            return default
        status = "unresolved"
        return match.group(0)

    return _PLACEHOLDER_RE.sub(repl, raw), status


def normalize_path(path: str) -> str:
    text = strip_quotes(str(path or "").strip())
    if not text:
        return "/"
    if not text.startswith("/"):
        text = "/" + text
    text = re.sub(r"/+", "/", text)
    return text.rstrip("/") or "/"


def combine_paths(prefixes: Sequence[str], suffixes: Sequence[str]) -> List[str]:
    base = list(prefixes) or [""]
    tail = list(suffixes) or [""]
    out: List[str] = []
    for left in base:
        for right in tail:
            out.append(normalize_path("/".join([left.strip("/"), right.strip("/")]).strip("/")))
    return sorted(set(out))


def _split_top_level(text: str) -> List[str]:
    items: List[str] = []
    current: List[str] = []
    depth = 0
    quote = ""
    escaped = False
    for char in text:
        if escaped:
            current.append(char)
            escaped = False
            continue
        if char == "\\":
            current.append(char)
            escaped = True
            continue
        if quote:
            current.append(char)
            if char == quote:
                quote = ""
            continue
        if char in {"'", '"'}:
            quote = char
            current.append(char)
            continue
        if char in "({[":
            depth += 1
        elif char in ")}]":
            depth = max(0, depth - 1)
        if char == "," and depth == 0:
            item = "".join(current).strip()
            if item:
                items.append(item)
            current = []
            continue
        current.append(char)
    item = "".join(current).strip()
    if item:
        items.append(item)
    return items

File: rust/crates/analyzer-jvm-overlays/src/spring/value_resolver.rs
//! Port `tools/spring/value_resolver.py` — annotation args parsing, path
//! normalize/combine, placeholder resolve.

use std::collections::BTreeMap;

use regex::Regex;
use serde_json::{Value, Value as Json};

use crate::pyjson::py_str;

pub struct ValueResolver {
    string_re: Regex,
    named_arg_re: Regex,
    placeholder_re: Regex,
}

impl Default for ValueResolver {
    fn default() -> Self {
        Self::new()
    }
}

impl ValueResolver {
    pub fn new() -> Self {
        Self {
            string_re: Regex::new(r#""([^"\\]*(?:\\.[^"\\]*)*)"|'([^'\\]*(?:\\.[^'\\]*)*)'"#).unwrap(),
            named_arg_re: Regex::new(r"([A-Za-z_][\w.-]*)\s*=\s*([^,]+(?:\{[^}]*\})?)").unwrap(),
            placeholder_re: Regex::new(r"\$\{([^}:]+)(?::([^}]+))?\}").unwrap(),
        }
    }

    pub fn strip_quotes(&self, value: &str) -> String {
        let text = value.trim();
        let bytes = text.as_bytes();
        if bytes.len() >= 2 && bytes[0] == bytes[text.len() - 1] && (bytes[0] == b'\'' || bytes[0] == b'"') {
            return text[1..text.len() - 1].to_string();
        }
        text.to_string()
    }

    pub fn extract_string_literals(&self, text: &str) -> Vec<String> {
        let mut values = Vec::new();
        for capture in self.string_re.captures_iter(text) {
            if let Some(group) = capture.get(1) {
                values.push(group.as_str().to_string());
            } else if let Some(group) = capture.get(2) {
                values.push(group.as_str().to_string());
            }
        }
        values
    }

    /// `parse_annotation_args`.
    pub fn parse_annotation_args(&self, raw_args: &str) -> BTreeMap<String, Json> {
        let mut text = raw_args.trim().to_string();
        if text.starts_with('(') && text.ends_with(')') && text.len() >= 2 {
            text = text[1..text.len() - 1].trim().to_string();
        }
        let mut result: BTreeMap<String, Json> = BTreeMap::new();
        if text.is_empty() {
            return result;
        }
        for capture in self.named_arg_re.captures_iter(&text) {
            let key = capture.get(1).unwrap().as_str().to_string();
            let value = capture.get(2).unwrap().as_str().to_string();
            result.insert(key, self.parse_value(&value));
        }
        if result.is_empty() {
            result.insert("value".to_string(), self.parse_value(&text));
        }
        result
    }

    /// `parse_value`.
    pub fn parse_value(&self, value: &str) -> Json {
        let text = value.trim();
        if text.starts_with('{') && text.ends_with('}') && text.len() >= 2 {
            let inner = &text[1..text.len() - 1];
            return Json::Array(
                split_top_level(inner)
                    .iter()
                    .map(|item| self.parse_value(item))
                    .collect(),
            );
        }
        let strings = self.extract_string_literals(text);
        if strings.len() == 1 && (text == format!("\"{}\"", strings[0]) || text == format!("'{}'", strings[0])) {
            return Json::String(strings[0].clone());
        }
        if !strings.is_empty() && text.starts_with('{') {
            return Json::Array(strings.into_iter().map(Json::String).collect());
        }
        let lower = text.to_lowercase();
        if lower == "true" || lower == "false" {
            return Json::Bool(lower == "true");
        }
        Json::String(self.strip_quotes(text))
    }

    /// `first_arg` — giá trị đầu có mặt trong args (không kiểm tra truthy).
    pub fn first_arg(&self, args: &BTreeMap<String, Json>, names: &[&str], default: &str) -> Json {
        for name in names {
            if let Some(value) = args.get(*name) {
                return value.clone();
            }
        }
        Json::String(default.to_string())
    }

    /// `list_arg`.
    pub fn list_arg(&self, args: &BTreeMap<String, Json>, names: &[&str]) -> Vec<String> {
        let value = self.first_arg(args, names, "");
        if value.is_null() || matches!(&value, Json::String(s) if s.is_empty()) {
            return Vec::new();
        }
        match &value {
            Json::Array(items) => items
                .iter()
                .map(py_str)
                .filter(|item| !item.is_empty())
                .collect(),
            other => vec![py_str(other)],
        }
    }

    /// `resolve_placeholders` — trả (resolved, status).
    pub fn resolve_placeholders(&self, value: &str, config_index: &BTreeMap<String, Vec<String>>) -> (String, String) {
        let raw = value.to_string();
        let mut status = "resolved".to_string();
        let replaced = self
            .placeholder_re
            .replace_all(&raw, |captures: &regex::Captures| {
                let key = captures.get(1).map(|m| m.as_str()).unwrap_or("");
                let default = captures.get(2).map(|m| m.as_str());
                let candidates = config_index.get(key).cloned().unwrap_or_default();
                if let Some(last) = candidates.last() {
                    return last.clone();
                }
                if let Some(default) = default {
                    status = "defaulted".to_string();
                    return default.to_string();
                }
                status = "unresolved".to_string();
                captures.get(0).unwrap().as_str().to_string()
            })
            .to_string();
        (replaced, status)
    }

    /// `normalize_path`.
    pub fn normalize_path(&self, path: &str) -> String {
        let mut text = self.strip_quotes(path.trim());
        if text.is_empty() {
            return "/".to_string();
        }
        if !text.starts_with('/') {
            text = format!("/{text}");
        }
        while text.contains("//") {
            text = text.replace("//", "/");
        }
        let trimmed = text.trim_end_matches('/').to_string();
        if trimmed.is_empty() {
            "/".to_string()
        } else {
            trimmed
        }
    }

    /// `combine_paths` — product + sorted unique.
    pub fn combine_paths(&self, prefixes: &[String], suffixes: &[String]) -> Vec<String> {
        let base: Vec<String> = if prefixes.is_empty() {
            vec![String::new()]
        } else {
            prefixes.to_vec()
        };
        let tail: Vec<String> = if suffixes.is_empty() {
            vec![String::new()]
        } else {
            suffixes.to_vec()
        };
        let mut out: Vec<String> = Vec::new();
        for left in &base {
            for right in &tail {
                let joined = format!("{}/{}", left.trim_matches('/'), right.trim_matches('/'));
                let stripped = joined.trim_matches('/').to_string();
                out.push(self.normalize_path(&stripped));
            }
        }
        out.sort();
        out.dedup();
        out
    }
}

/// `_split_top_level` của value_resolver (depth tính `({[<`).
pub fn split_top_level(text: &str) -> Vec<String> {
    let mut items: Vec<String> = Vec::new();
    let mut current = String::new();
    let mut depth = 0i32;
    let mut quote: Option<char> = None;
    let mut escaped = false;
    for ch in text.chars() {
        if escaped {
            current.push(ch);
            escaped = false;
            continue;
        }
        match quote {
            Some(q) => {
                current.push(ch);
                if ch == q {
                    quote = None;
                }
            }
            _ => match ch {
                '\\' => {
                    current.push(ch);
                    escaped = true;
                }
                '\'' | '"' => {
                    quote = Some(ch);
                    current.push(ch);
                }
                '(' | '{' | '[' | '<' => {
                    depth += 1;
                    current.push(ch);
                }
                ')' | '}' | ']' | '>' => {
                    depth = (depth - 1).max(0);
                    current.push(ch);
                }
                ',' if depth == 0 => {
                    let item = current.trim().to_string();
                    if !item.is_empty() {
                        items.push(item);
                    }
                    current.clear();
                }
                _ => current.push(ch),
            },
        }
    }
    let item = current.trim().to_string();
    if !item.is_empty() {
        items.push(item);
    }
    items
}

/// Wrapper tiện dụng — mirror hàm module-level của Python.
pub fn parse_annotation_args(raw_args: &str) -> BTreeMap<String, Value> {
    ValueResolver::new().parse_annotation_args(raw_args)
}
